### ISSUE TITLE
Best number of epochs

### ISSUE THREAD

**Issue by JulienGervai** [sdv-dev/SDV#222]
Hello,

I got a question concerning the best number of epochs to train CTGAN or CopulaGAN. I see that in most cases the Loss functions will decrease and if you train as much it will be a negative number. 
So my question is, do we want the loss function to be very very low (like a very little negative number) or do we want it to be close to 0 ?
And, can we determine the best number of epochs just by looking at the loss functions ?
Can the model be overfitted even if we have 2 neural networks ?

Thank you

**Comment by csala** [comment]
Hello @JulienGervai 

Those are really interesting questions! In general, the Early Stopping criteria for GANs is a tricky debate, as the loss values by themselves usually do not provide enough information. At the end, when the loss values are computed the performance of both networks contribute in opposite directions, which means that a stable loss could simply indicate that both networks are learning at the same rate, but still improving.

So, even though it would be interesting to have an Early Stopping mechanism that allows unnecessary epochs, this is something that should not rely only on the loss values, but rather on other metrics that should be added specifically for it (like a machine learning efficacy based early stopping). And something which we can definitely explore!

**Comment by JulienGervai** [comment]
Thanks for your answer, 

I was curious because I saw that you have already answer a question on this subject (for CTGAN and not for SDV) https://github.com/sdv-dev/CTGAN/issues/66 . 
And to answer this question in your answer you said : "In most cases, a lower number of epochs, of just a few dozens, can be more than enough to explore a particular problem a bit faster and get an idea of what the model can do on your data. However, if you want to get the most out of the model, you will probably need to tweak it a little bit and find the optimal value for each dataset you work on."
Which is totally true. And I did some tests and for example with this dataset : https://www.kaggle.com/mirichoi0218/insurance
After spliting the data into two dataframe train and test, if you train the model on the dataframe train and generate some data with it.
You have someting like that with 3000 epochs and a batch size equal to 100 : 

![splitting](https://user-images.githubusercontent.com/69855022/97544059-5f203800-19c9-11eb-8ad6-2f95f84b0f33.PNG)
![training](https://user-images.githubusercontent.com/69855022/97544145-78c17f80-19c9-11eb-97bf-fe8e03c72eb2.PNG)
We can see that the distribution for charges is well learnt
![distribution](https://user-images.githubusercontent.com/69855022/97544152-7b23d980-19c9-11eb-9c67-b3ffe4bb0cc1.PNG)
And the model have understood the correlation between smoker and charges 
![correlation](https://user-images.githubusercontent.com/69855022/97544235-98f13e80-19c9-11eb-8940-44f77a3264ba.PNG)
If we look at a ML test : 
![testML](https://user-images.githubusercontent.com/69855022/97544287-aad2e180-19c9-11eb-8792-fe19722bc42c.PNG)
The result with the fake(second) data is close to the real data(first)



However, if I do the same thing with the default value 300 epochs I have : 



![distrib300](https://user-images.githubusercontent.com/69855022/97544865-7c093b00-19ca-11eb-993d-472400b09f61.PNG)
![corr300](https://user-images.githubusercontent.com/69855022/97544868-7dd2fe80-19ca-11eb-9b62-4858ead190f2.PNG)
![ML300](https://user-images.githubusercontent.com/69855022/97544876-7f9cc200-19ca-11eb-8eb0-90272357460c.PNG)


That is why I am worried about overfitting because I have to set the number of epochs to a very high number.
